contracts/TokenCrowdsale.sol

```diff
@@ -1,13 +1,183 @@
+
+    // while we can purchase all tokens in current cap-rate step, move to other step
+    while (_weiAmount > 0 && _currentSupply < capsTo[capsTo.length.sub(1)]) {
+      _tokensForRate = _capTo.sub(_currentSupply);
+      _weiReq = _tokensForRate.div(_rate);
+      if (_weiReq < _weiAmount) {
+        // if wei required for tokens is less than we have - all tokens from the step can be pruchased
+         _tokenAmount = _tokenAmount.add(_tokensForRate);
+      } else {
+        // if wei required is more or equal than we have - we can purchase only part of the cap-rate step tokens
+         _tokenAmount = _tokenAmount.add(_weiAmount.mul(_rate));
+         _weiReq = _weiAmount;
+      }
+
+      _weiSpent = _weiSpent.add(_weiReq);
+      _weiAmount = _weiAmount.sub(_weiReq);
+      _currentSupply = token.totalSupply().add(_tokenAmount);
+      (_capTo, _rate) = _getRateFor(_currentSupply);
+    }
+
+    super._processPurchase(_beneficiary, _tokenAmount);
+    _processFundsOverflow(_beneficiary, _weiSpent);
+  }
+
+  /*
+    Method handles uncapped state purchase. It is needed to check if beneficiary
+    sent more ethers than there are tokens to purchase.
+  */
+  function _processUncappedPurchase(address _beneficiary, uint256 _tokenAmount) internal {
+    uint256 _currentSupply = token.totalSupply();
+    if (_currentSupply.add(_tokenAmount) > icoTotalCap) {
+      _tokenAmount = icoTotalCap.sub(_currentSupply);
+    }
+    super._processPurchase(_beneficiary, _tokenAmount);
+    uint256 _weiAmount = _tokenAmount.div(rate);
+    _processFundsOverflow(_beneficiary, _weiAmount);
+  }
+
+  /*
+    In either stage last beneficiary can send more ethers than there are tokens
+    to purchase - in this case overflow ethers are returned to beneficiary
+  */
+  function _processFundsOverflow(address _beneficiary, uint256 _weiAmount) internal {
+    require(_weiAmount <= msg.value);
+    if (msg.value > _weiAmount) {
+      uint256 _weiToReturn = msg.value.sub(_weiAmount);
+      weiRaised = weiRaised.sub(_weiAmount);
+      _beneficiary.transfer(_weiAmount);
+    }
+  }
+
+  /*
+    Method for retrieving rate and capTo for provided supply. It is used in capped
+    state were rate depends on sold tokens amount
+  */
+  function _getRateFor(uint256 _supply) internal view returns(uint256, uint256) {
+    require(capsTo[capsTo.length - 1] > _supply);
+    uint256 _i = 0;
+    while(_i < capsTo.length) {
+      if (capsTo[_i] > _supply) {
+        return (capsTo[_i], rates[_i]);
+      }
+      _i = _i.add(1);
+    }
+  }
+
+  // OpenZeppelin FinalizableCrowdsale method override
+  function finalization() internal {
+    // allow finalize only once
+    require(finalizedTime == 0);
+
+    Token _token = Token(token);
+
+    require(_token.mint(foundationWallet, _token.cap().mul(foundationPercentage).div(100)));
+
+    uint256 _leftoverTokens = icoTotalCap.sub(_token.totalSupply());
+    require(_token.mint(lockupWallet, _leftoverTokens));
```

Comment:
I think you should do this first - before minting fundation tokens. 
If you mint fundation tokens first totalSupply will be: fundationTokens + allSold which might even be more than icoTotalCap. 

Or if you want to do it here - you should use _token.cap() instead of a icoTotalCap, since all that is left should be assigned to lockup anyway.